src/panels/lovelace/cards/hui-thermostat-card.ts

```diff
@@ -494,11 +499,37 @@ export class HuiThermostatCard extends LitElement implements LovelaceCard {
     });
   }
 
+  protected firstUpdated(): void {
+    if (!this.validatePreviousMode) {
+      this.validatePreviousMode = debounce(
+        (previousStateObj: ClimateEntity) => {
+          const stateObj = this.hass!.states[
+            this._config!.entity
+          ] as ClimateEntity;
+          if (previousStateObj === stateObj && this.previousMode) {
+            stateObj.state = this.previousMode;
+            this.previousMode = undefined;
+            this.requestUpdate();
+          }
+        },
+        2000
+      );
+    }
+  }
+
   private _handleAction(e: MouseEvent): void {
+    const stateObj = this.hass!.states[this._config!.entity] as ClimateEntity;
+    const selectedMode = (e.currentTarget as any).mode;
+
+    this.previousMode = this.previousMode || stateObj.state;
+    stateObj.state = selectedMode;
```

Comment:
we should not update the `stateObj` instead make a property that stores the optimistic value and use that during rendering. In the update call we check if `stateObj` is updated and update the optimistic value.

Reply:
See: https://github.com/home-assistant/frontend/blob/dev/src/components/entity/ha-entity-toggle.ts#L141-L154

Reply:
Thanks I had a feeling that it probably shouldn't be done. Will update

Reply:
Updated.

Reply:
I now remember why I did it this way. If only my memory wasn't so foggy earlier. 

Its because the `stateColorCss` method takes in a stateObj which I cannot provide unless I provide a deep copy of the existing stateObj. It seems like there would either have to be a hack involved or a refactor of this method to enable it to provide colors from a string or input other than the stateObj. I went ahead and implemented it this way, but its not very pretty. Not sure how to workaround this or if there is a better alternative.